### PR TITLE
Dockerization

### PR DESCRIPTION
**Pull request by bindeali**
## Info on update bcb5675:
* Added GitHub action for publishing the Docker image.
* Changed validation URL to the env variable.
* (updated env variables in deploy, so this [deploy preview](https://deploy-preview-235--ontographer.netlify.app/) should start ontoGrapher)

## Info on update 3ddfccc:
* Added YAML config file support.
* Changed the Components type to fully conform to the official design.
* Changed the ENV loading to support loading through `window.__config__`.
* Changed the Dockerfile to use the desired nginx configuration.
* Original description of the PR has been removed as it is no longer accurate.

### CODE REVIEW

**Review by karelklima — COMMENTED** [review]
Some things to discuss - especially the auth realm and how it ties to Keycloak.

Overall the PR looks good. The only thing I'd suggest is to configure your IDE or use some linter / code formatter that would attach newlines at the end of each file (see https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline)

**Review by psiotwo — DISMISSED** [review]
lgtm

**Review by karelklima — CHANGES_REQUESTED** [review]
Correct docker build path and circular dependencies (Variables.ts) need to be resolved before merging.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM node:12 as node
+
+ARG PUBLIC_PATH=/
+ENV PUBLIC_URL=${PUBLIC_PATH}
+
+WORKDIR /ontographer
+COPY package.json .
+RUN npm install --production
+COPY . .
+RUN npm run build
```
[the remark this answers is not in the document]
Comment (bindeali) [reply Dockerfile:11]: Added an eslint check in addition to the check.

File: src/config/Variables.ts [hunk 2]
```diff
@@ -4,6 +4,8 @@ import {initLanguageObject} from "../function/FunctionEditVars";
 import {Restriction} from "../datatypes/Restriction";
 import {Representation} from "./Enum";
 import * as joint from "jointjs";
+import {Components} from "../datatypes/Components";
+import {getComponentsVariable, getEnvironmentVariable} from "../function/FunctionGetVars";
```
[the remark this answers is not in the document]
Comment (bindeali) [reply src/config/Variables.ts]: Separated the environment module(s) into Environment.ts and removed the circular dependency.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,13 @@
+FROM node:12 as node
+
+ARG PUBLIC_PATH=/
+ENV PUBLIC_URL=${PUBLIC_PATH}
```
[the remark this answers is not in the document]
Comment (bindeali) [reply Dockerfile:4]: Changed the PUBLIC_PATH to "."

File: .github/workflows/docker-publish.yml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    runs-on: ubuntu-latest
+    if: github.event_name == 'push'
+
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Build image
+        run: docker build . --file Dockerfile --tag $IMAGE_NAME
+
+      - name: Log into registry
```
[the remark this answers is not in the document]
Comment (bindeali) [reply .github/workflows/docker-publish.yml:24]: I have swapped the commands as instructed.

**Review by karelklima — APPROVED** [review]
LGTM

**Pull request merged**